### PR TITLE
Deterministic ports for grpc/http api servers

### LINKED ISSUE #6375: `--rpc-port` should either succeed or fail (TestRunPortForward flake)
[`TestRunPortForward`](https://github.com/GoogleContainerTools/skaffold/blob/main/integration/port_forward_test.go#L55) flakes on Travis on occasion.  The problem is that we run Skaffold requesting a specific RPC port (35964) but it is seemingly unavailable and so Skaffold chooses a different port instead of failing.  The test waits trying to connect to the requested RPC port and times out.

Basically we should honour the requested port specified in `--rpc-port` and fail if unavailable.  This issue is somewhat similar to the reasoning in #5839.

```
=== RUN   TestRunPortForward
time="2021-08-05T01:07:28Z" level=trace msg="Partitions: "
time="2021-08-05T01:07:28Z" level=trace msg="P0 total: 300.000000, size: 6\n"
time="2021-08-05T01:07:28Z" level=trace msg="P1 total: 299.980000, size: 8\n"
time="2021-08-05T01:07:28Z" level=trace msg="P2 total: 299.990000, size: 15\n"
time="2021-08-05T01:07:28Z" level=trace msg="P3 total: 87.830000, size: 16\n"
time="2021-08-05T01:07:28Z" level=debug msg="getting client config for kubeContext: ``"
time="2021-08-05T01:07:28Z" level=info msg="Namespace: skaffoldc2wlk"
time="2021-08-05T01:07:28Z" level=info msg="Running [skaffold run --namespace skaffoldc2wlk --default-repo gcr.io/k8s-skaffold --port-forward --rpc-port 35964 --enable-rpc] in examples/microservices"
    rpc_test.go:378: waiting for connection...
time="2021-08-05T01:07:28Z" level=warning msg="starting gRPC server on port 35965. (35964 is already in use)"
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:378: waiting for connection...
    rpc_test.go:380: error retrieving event log: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp :35964: connect: connection refused"
    helper.go:194: Skaffold log:
         Generating tags...
        >  - leeroy-web -> gcr.io/k8s-skaffold/leeroy-web:v1.29.0-12-g2ffaeeb
        >  - leeroy-app -> gcr.io/k8s-skaffold/leeroy-app:v1.29.0-12-g2ffaeeb
        >  - base -> gcr.io/k8s-skaffold/base:v1.29.0-12-g2ffaeeb
        > Checking cache...
        >  - leeroy-web: Found Locally
        >  - leeroy-app: Found Locally
        >  - base: Found Locally
        > Starting test...
        > Tags used in deployment:
        >  - leeroy-web -> gcr.io/k8s-skaffold/leeroy-web:81bea89013cde47d2d10613a4efc4389d18fca03dbc35e7b85c90c8b010fc408
        >  - leeroy-app -> gcr.io/k8s-skaffold/leeroy-app:3963b2bcd69186ca063a4a38e1c69b6fa7b1f3774e8991a4cb7cb4db60a78533
        >  - base -> gcr.io/k8s-skaffold/base:285e9d36fd581f2e88fc59983275583b49d7d6e73bb6e1572f99f97306ef9263
        > Starting deploy...
        > Loading images into k3d cluster nodes...
        >  - gcr.io/k8s-skaffold/leeroy-web:81bea89013cde47d2d10613a4efc4389d18fca03dbc35e7b85c90c8b010fc408 -> Found
        >  - gcr.io/k8s-skaffold/leeroy-app:3963b2bcd69186ca063a4a38e1c69b6fa7b1f3774e8991a4cb7cb4db60a78533 -> Found
        > Images loaded in 132.101079ms
        >  - deployment.apps/leeroy-web created
        >  - service/leeroy-app created
        >  - deployment.apps/leeroy-app created
        > Waiting for deployments to stabilize...
        >  - skaffoldc2wlk:deployment/leeroy-web is ready. [1/2 deployment(s) still pending]
        >  - skaffoldc2wlk:deployment/leeroy-app is ready.
        > Deployments stabilized in 2.402 seconds
        > Press Ctrl+C to exit
        > Port forwarding deployment/leeroy-web in namespace skaffoldc2wlk, remote port 8080 -> http://127.0.0.1:9002
        > Port forwarding service/leeroy-app in namespace skaffoldc2wlk, remote port 50051 -> http://127.0.0.1:50053
        > Port forwarding deployment/leeroy-app in namespace skaffoldc2wlk, remote port http -> http://127.0.0.1:9003
        > 
time="2021-08-05T01:07:48Z" level=info msg="Ran [skaffold run --namespace skaffoldc2wlk --default-repo gcr.io/k8s-skaffold --port-forward --rpc-port 35964 --enable-rpc] in 20.106 seconds"
--- FAIL: TestRunPortForward (20.12s)
```

### LINKED ISSUE #6417: [Flake] TestEventsRPC is flaky
I have seen this multiple times 
```
rpc_test.go:93: error retrieving event log: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial tcp :36288: connect: connection refused"
    helper.go:194: Skaffold log:
         Listing files to watch...
        >  - test-dev
        > Generating tags...
        >  - test-dev -> gcr.io/k8s-skaffold/test-dev:v1.29.0-35-gba4e2a8-dirty
        > Starting build...
        > Found [kind-kind] context, using local docker daemon.
        > Building [test-dev]...
        > Sending build context to Docker daemon  3.072kB
        > Step 1/3 : FROM busybox
        >  ---> 69593048aa3a
        > Step 2/3 : COPY foo /foo
        >  ---> Using cache
        >  ---> 608862fe6e54
        > Step 3/3 : CMD while true; do cat /foo; sleep 1; done
        >  ---> Using cache
        >  ---> c302772802f6
        > Successfully built c302772802f6
        > Successfully tagged gcr.io/k8s-skaffold/test-dev:v1.29.0-35-gba4e2a8-dirty
        > Starting test...
        > Tags used in deployment:
        >  - test-dev -> gcr.io/k8s-skaffold/test-dev:c302772802f6c4ce0c7d4d200f03fdcd0cdac4ac74ae980763254fdd55ebb205
        > Starting deploy...
        > Loading images into kind cluster nodes...
        >  - gcr.io/k8s-skaffold/test-dev:c302772802f6c4ce0c7d4d200f03fdcd0cdac4ac74ae980763254fdd55ebb205 -> Found
        > Images loaded in 194.78357ms
        >  - deployment.apps/test-dev created
        > Press Ctrl+C to exit
        > Watching for changes...
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > [test-dev]foo
        > 
--- FAIL: TestEventsRPC (20.84s)
```

### LINKED ISSUE #6425: Proposal: strict port handling and enable flags for Skaffold API (deprecate --enable-rpc)
## Problem

Right now our GRPC/HTTP servers try to choose a port that's different than the provided port number if that's busy. This is problematic for a number of reasons:

1. we query the system to find `usedPorts`, and then make a decision based on that → this is **not atomic** (something might've bind()ed to that port by then, which makes the tests flaky)
2. potential callers of skaffold who provide a preferred port would not<sup>[citation needed]</sup>  expect skaffold to choose a different port (or they will be told to parse `starting gRPC server on port %d`) → **either case bad UX fo the caller?**
3. it's already causing test flakes: #6375 #6417

Context: I believe this port choosing logic was added in response to #4253.

## Proposed Solution

1. **Do not start API server by default** unless the caller specifies `--rpc-port` or both `--rpc-http-port`. (Current behavior: by default `dev`/`run` commands start grpc/http ports 50051/50052, and find adjacent available ports if they are not available).

1. Usage of either `--rpc-port` or `--rpc-http-port` now imply `--enable-rpc=true`.

   :warning: This means the **`--enable-rpc` flag will be deprecated.** All notable customers such as Cloud Code for VSC+IntelliJ are impacted from this.
   
   :warning: This flag will be marked as deprecated in the v1.31 release, and will be removed in v1.32 or in 3 months (whichever is longer, per [deprecation policy](https://skaffold.dev/docs/references/deprecation/)) since the API is a beta feature. 

1. If specified port numbers are busy, **Skaffold now fails instead of trying to find an adjacent available port.** (Let caller find a free port, they can call random() if they like). We might 

## Impact on Existing Users

**Taken only Cloud Code into consideration:**

This is a breaking change for those (1) relying on API servers starting by default and (2) those relying on skaffold's "find an adjacent open port if given port is unavailable" (and parse the actual port from logs). Thankfully Cloud Code is unimpacted:

* CC for VSCode uses only `--rpc-http-port` and they use [a library](https://github.com/sindresorhus/get-port) to find an available port (which uses :0 mechanism, more prone to collisions). They also use `--enable-rpc=true` which will be deprecated with this proposal. 
* CC for IntelliJ uses only `--rpc-port` and they also scan local ports to find an open port (chooses rand port from an open range, less prone to collisions.) They also use `--enable-rpc=true` which will be deprecated with this proposal.

In both cases, the port finding is susceptible to a race condition where the caller thinks the found port number is available, and they pass it to Skaffold, but some other process might've started using it. I don't think we can fix this problem (short of using Unix domain sockets and Windows named pipes).

/cc @nkubala 
/area cleanup
/kind design discussion

### PR DESCRIPTION
**Pull request by ahmetb**
Fixes #6425
Fixes #6417
Fixes #6375

**Description**

skaffold tries to find as different available port
than the specified port. As outlined in #6425,
this poses several problems such as not being api
to atomically secure an available port,
redundant internal bookkeeping of port numbers
that are allocated (unused code), and making life
harder for API callers as they need to parse
unstructured logs to find out the port numbers.

This patch makes skaffold more strict about how
port numbers for its internal APIs are bind()ed.

This patch also fixes test flakes for the following
integration tests (each ran 100 times): 

- TestEventsRPC #6417 
- TestRunPortForward #6375

**User facing changes**

This patch changes the visible behavior of API servers
(for users of skaffold API):

* no more default ports 50051 and 50052 for APIs
  (they are decided randomly by default)

* if given port is busy, we fail with error so the
  caller can specify another port (TODO we need
  to discuss if we want a custom exitcode to
  signal this)

**Follow-up Work (remove if N/A)**

Determine if we need a custom exitcode indicating
"port you specified is not available" for callers
of the API (e.g. Cloud Code).

### CODE REVIEW

File: pkg/skaffold/constants/constants.go [hunk 1]
```diff
@@ -61,8 +61,8 @@ const (
 	DefaultCacheFile   = "cache"
 	DefaultMetricFile  = "metrics"
 
-	DefaultRPCPort     = 50051
-	DefaultRPCHTTPPort = 50052
+	DefaultRPCPort     = 0 // choose an available port at runtime
```
Comment (gsquared94) [inline pkg/skaffold/constants/constants.go:64]: would changing the default port from 50051/52 to random be considered as breaking backwards compatibility?
IMO there could be users with scripts or other code that assume these port numbers. Could we not continue trying to assign these ports, but fallback to random only if unavailable?

Comment (ahmetb) [reply pkg/skaffold/constants/constants.go:64]: > fallback to random only if unavailable?

that's sort of the problem described in #6425.
silently picking ports will probably life harder for existing integrations.

> would changing the default port from 50051/52 to random be considered as breaking backwards compatibility?

Yes. I opened this per offline chat, but this definitely warrants discussion if we want to do this or not. My assumption was that there are no notable users of these APIs yet but I lack enough context to run with that assumption.

**Comment by ahmetb** [comment]
Maybe the right approach is to not start any ports on run/dev commands unless explicitly requested (right now we do). That way, ports are opened only if skaffold is started by a high-level IDE etc needing these ports to be open.

**Comment by ahmetb** [comment]
I'll close this and open a fresh one when it's not WIP.

**Pull request closed without merging**